### ISSUE TITLE
Amazon Dash Button binding

### ISSUE THREAD

**Issue by OLibutzki** [openhab/openhab2-addons#1235]
The Amazon Dash Button is a very cheap device which can be integrated easily into openHAB. There are solutions like Dasher (https://github.com/maddox/dasher), but they rely on NodeJS and you need some extra unnecessary indirections.

So I prefer a native openHAB binding.
##

**Comment by kaikreuzer** [comment]
Meaning you are about to contribute one...?

**Comment by OLibutzki** [comment]
Wait a second... ;-)

**Comment by kaikreuzer** [comment]
Ah 👍

**Comment by OLibutzki** [comment]
I'm pretty sure you already have a Dash button, don't you? So testing should be straightforward :-)

**Comment by kaikreuzer** [comment]
No, I don't! Please be patient with my review anyhow. I am about to leave for JavaOne and I have a long list of bindings to still review. But if you are bored in the meantime: How about helping to review some other binding, so I will be quicker in getting to yours? :-)

**Comment by OLibutzki** [comment]
Haha, honestly, I'm happy that I found the time to implement this binding.

Take your time. I can use the binding, so I do not care that much how long it takes to merge it.

**Comment by dominicdesu** [comment]
@OLibutzki is it necessary to run OH2 as root in order for this to work?
I started the discovery process but cannot detect the dash button, even though I can see an ARP package in Wireshark.
Only log I can see in OH2 is this:
```
2016-09-20 22:43:08.137 [INFO ] [org.pcap4j.core.Pcaps               ] - No NIF was found.
2016-09-20 22:43:09.174 [INFO ] [org.pcap4j.core.Pcaps               ] - No NIF was found.
2016-09-20 22:43:09.225 [INFO ] [org.pcap4j.core.Pcaps               ] - No NIF was found.
```

Any idea?

**Comment by dominicdesu** [comment]
Ah just find it on the [Pcap4J website](https://github.com/kaitoy/pcap4j#others): 

> Pcap4J needs administrator/root privileges. Or, if on Linux, you can run Pcap4J with a non-root user by granting capabilities CAP_NET_RAW and CAP_NET_ADMIN to your java command by the following command: setcap cap_net_raw,cap_net_admin=eip /path/to/java

**Comment by OLibutzki** [comment]
@dominicdesu I updated the documentation.

Does the binding work for you?

**Comment by dominicdesu** [comment]
No really, the OH2 process just crashes without any error if I start the discovery:
```
2016-09-21 21:24:52.731 [WARN ] [org.pcap4j.util.PropertiesLoader    ] - [org/pcap4j/pcap4j.properties] Could not get value by org.pcap4j.af.inet, use default value: null
2016-09-21 21:24:52.733 [WARN ] [org.pcap4j.util.PropertiesLoader    ] - [org/pcap4j/pcap4j.properties] Could not get value by org.pcap4j.af.inet6, use default value: null
2016-09-21 21:24:52.735 [WARN ] [org.pcap4j.util.PropertiesLoader    ] - [org/pcap4j/pcap4j.properties] Could not get value by org.pcap4j.af.packet, use default value: null
2016-09-21 21:24:52.736 [WARN ] [org.pcap4j.util.PropertiesLoader    ] - [org/pcap4j/pcap4j.properties] Could not get value by org.pcap4j.af.link, use default value: null
2016-09-21 21:24:52.788 [INFO ] [org.pcap4j.core.Pcaps               ] - 5 NIF(s) found.
2016-09-21 21:24:53.744 [INFO ] [org.pcap4j.core.Pcaps               ] - 5 NIF(s) found.
2016-09-21 21:26:15.568 [INFO ] [org.pcap4j.core.Pcaps               ] - 5 NIF(s) found.
2016-09-21 21:26:15.680 [WARN ] [org.pcap4j.util.PropertiesLoader    ] - [org/pcap4j/pcap4j.properties] Could not get value by org.pcap4j.dlt.raw, use default value: null
2016-09-21 21:26:15.703 [INFO ] [org.pcap4j.core.PcapHandle          ] - Starting loop.
2016-09-21 21:26:15.847 [INFO ] [org.pcap4j.core.PcapHandle          ] - Starting loop.
```

Any idea? I assume some issue with the native library, but I don't know where to look for a error log or similar... :-(

**Comment by OLibutzki** [comment]
I have only tested with Windows so far. I wonder why there are 5 NIFs found and only two loops are started.

I guess I have to test with Linux as well.

**Comment by dominicdesu** [comment]
ifconfig shows 3 network interfaces: wlan, ethernet and local loopback.
I added some logging to your binding which shows the following before crashing:
```
2016-09-21 21:09:21.113 [INFO ] [org.pcap4j.core.Pcaps               ] - 5 NIF(s) found.
2016-09-21 21:09:21.117 [INFO ] [ery.AmazonDashButtonDiscoveryService] - Listening on device name: [eth0] description: [null] address: [/192.168.178.51] loopBack: [false]] local: [true]
2016-09-21 21:09:21.211 [WARN ] [org.pcap4j.util.PropertiesLoader    ] - [org/pcap4j/pcap4j.properties] Could not get value by org.pcap4j.dlt.raw, use default value: null
2016-09-21 21:09:21.224 [INFO ] [ery.AmazonDashButtonDiscoveryService] - Listening on device name: [wlan0] description: [null] loopBack: [false]] local: [true]
2016-09-21 21:09:21.231 [INFO ] [org.pcap4j.core.PcapHandle          ] - Starting loop.
2016-09-21 21:09:21.296 [INFO ] [ery.AmazonDashButtonDiscoveryService] - Listening on device name: [nflog] description: [Linux netfilter log (NFLOG) interface] loopBack: [false]] local: [true]
2016-09-21 21:09:21.296 [INFO ] [org.pcap4j.core.PcapHandle          ] - Starting loop.
```
So maybe it is crashing because it starts to looping on nflog?

I also tried out the dasher tool: Running without root doesn't work (`Error: No devices to listen`), while running with root works and detects my dash button correctly.

**Comment by dominicdesu** [comment]
Ha!
I added a Dash button thing manually (without discovery), inserted mac address and network interface, and it works!! :smile:

**Comment by OLibutzki** [comment]
Great! Can you check if an exception occurs after starting the loop? So I would catch this exception and ignore that network interface while scanning.

**Comment by dominicdesu** [comment]
I just checked, there is no exception thrown. The OH2 Java thread just terminates.

**Comment by OLibutzki** [comment]
@dominicdesu Thanks, I wonder if there is a criteria which I could use in order to determine if a network interface should be scanned...

**Comment by dominicdesu** [comment]
Hm, the dasher tool gives the following warning if I select nflog, but it doesn't crash:
`warning: SIOCGIFADDR: nflog: No such device - this may not actually work`

Btw. these are my 5 network interfaces:
```
1.eth0
2.wlan0
3.nflog (Linux netfilter log (NFLOG) interface)
4.any (Pseudo-device that captures on all interfaces)
5.lo
```

Maybe you can scan on "any", if such an NF exists on linux?

I have excluded the nflog interface from the capturing (comparing against `pcapNetworkInterface.getName()`) and now the discovery doesn't crash anymore.

**Comment by OLibutzki** [comment]
@dominicdesu I fixed a couple of bugs as you can see in #1236. So feel free to try the binding again.

**Comment by OLibutzki** [comment]
Closed the issue because #1236 is merged.
